### PR TITLE
Refactor models

### PR DESCRIPTION
**Pull request by AELSchauer**
So, this does a ton of refactoring on my API models and in my request service PORO to better follow  OOP from our lesson today. I haven't written any tests (I know! Bad Ashley!). Ali said this looked good, but I'm interested in your opinion, @s-espinosa.

Muchas gracias!

### CODE REVIEW

File: app/controllers/dashboard_controller.rb [hunk 1]
```diff
@@ -2,6 +2,6 @@ class DashboardController < ApplicationController
   before_action :authorize!
 
   def index
-    @user = current_user
+    @user = ViewUser.build(current_user.token, current_user.uid)
```
Comment (s-espinosa) [inline app/controllers/dashboard_controller.rb:5]: I probably wouldn't change it at this point, but I'm wondering if I would like `Redditor` better than ViewUser. I think this is fine, and I'm definitely being picky. Naming things is hard.

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 class PostsController < ApplicationController
-  def show
-    @post = Post.new(current_user.token, params[:title], params[:post_id])
+  def index
+    @post = Post.build(current_user.token, params[:subreddit_name], params[:post_id])
```
Comment (s-espinosa) [inline app/controllers/posts_controller.rb:3]: Curious as to why you moved from `.new` to `.build`. I don't have a strong opinion on it, so I'm just intrigued.

File: app/helpers/post_helper.rb [hunk 3]
```diff
@@ -1,17 +1,21 @@
...
-      ul_contents = ""
-      if comment.body.nil?
-        ul_contents << link_to(content_tag(:li, "#{comment.data["count"]} More Comments"), "#").html_safe
-      else
-        ul_contents << content_tag(:li, comment.body)
-      end
-      comment.replies.each do |reply|
-        ul_contents << recursive_list_tag(reply)
-      end
+      html = content_tag(:ul) {
```
Comment (s-espinosa) [inline app/helpers/post_helper.rb:3]: This is great! A lot going on here. Will think about other options, but you're totally making it work, and I think a helper is the right place for this kind of logic.

File: app/models/comment.rb [hunk 4]
```diff
@@ -1,18 +1,29 @@
...
-    if !@data["replies"].nil? && !@data["replies"].empty?
-      @data["replies"]["data"]["children"].each do |reply_hash|
-        replies << Comment.new(reply_hash, @depth+1)
+    if params["replies"].nil?
+      if not params["parent_id"].nil?
+        replies << Comment.new("id" => params["parent_id"], "body" => "")
+      end
+    elsif not params["replies"].empty?
+      params["replies"]["data"]["children"].map do |reply_hash|
+        replies << Comment.new(reply_hash["data"])
```
Comment (s-espinosa) [inline app/models/comment.rb:26]: 👍

File: app/services/reddit_request.rb [hunk 5]
```diff
@@ -4,7 +4,58 @@ def initialize(token)
...
+
+  def post_info(display_name, post_id)
+    response = request("/r/#{display_name}/comments/#{post_id}")
+    data = response[0]["data"]["children"][0]["data"]
+  end
+
+  def post_comments(display_name, post_id)
+    response = request("/r/#{display_name}/comments/#{post_id}")
+    response[1]["data"]["children"]
+  end
```
Comment (s-espinosa) [inline app/services/reddit_request.rb:60]: 🎉 Like this service.

File: app/views/shared/_user_info.html.erb [hunk 6]
```diff
@@ -2,11 +2,11 @@
 <table>
   <tr>
     <th>Link:</th>
-    <td><%= @user.karma[:link] %></td>
+    <td><%= @user.link_karma %></td>
```
Comment (s-espinosa) [inline app/views/shared/_user_info.html.erb:5]: I like this change!

**Review by s-espinosa — COMMENTED** [review]
Comments below. At a high level I like these changes. A couple of other notes:

- It seems like your routes file is getting a little wonky with pluralization/the route for viewing a post. I'm guessing there are reasons for those changes, but maybe we can talk about it?
- I spoke to Charlotte about an alternative pattern to passing the token around. Looking at your code I'm not positive I would make the same recommendation, but it's something to think about. Basically the idea is if these other pseudo models are related to User then you don't need to pass the token in the same way. I know that not all of these things (subreddits, posts, etc.) actually have a direct relationship with the user, so that might be a disadvantage. Let me know if you'd like to discuss.
- Recommend taking some time (maybe when you want a change of pace) to put together a README with instructions for setup. Also: noticed that you have your Reddit username hardcoded into your service. Recommend replacing that with another environment variable and then explaining in the README everything you'll need in your application.yml file. I've even seen something like an `example_application.yml` file that just has the basic outline of all the variables people will need to set. Seems like a nice-to-have to me.

Let me know if you have any specific questions! 🎉

File: app/controllers/dashboard_controller.rb (hunk 1 above)
Comment (AELSchauer) [reply app/controllers/dashboard_controller.rb:5]: That's so much better! Thanks!

File: app/controllers/posts_controller.rb (hunk 2 above)
Comment (AELSchauer) [reply app/controllers/posts_controller.rb:3]: Build allows me to make many request calls and pull in all the data I need in a specific view. In that build method, I can create an instance of that model that turns the data hash into instance variables.

**Pull request merged**